jigdump.1
.TH JIGDUMP 1 "November 2019" "Jigit jigdo tools"
.SH NAME
jigdump \- Dump the contents of a template file
.SH SYNOPSIS
.B jigdump
<template file>
.SH DESCRIPTION
.PP
jigdump will parse a template file and list its contents. Mainly
useful just for debugging jigdo and friends...
.SH "SEE ALSO"
\fBjigdo-file\fP(1), \fBjigit\fP(1), \fBjigsum\fP(1) and \fBmkimage\fP(1).
.SH "COPYRIGHT"
Copyright 2004-2019 Steve McIntyre (steve@einval.com)
.PP
jigdump may be copied under the terms and conditions of version 2
of the GNU General Public License, as published by the Free
Software Foundation (Cambridge, MA, USA).
.SH "AUTHOR"
Written by Steve McIntyre (steve@einval.com)


ChangeLog
Version 1.22 (2019-11-21)

 * Add support for SHA256 checksums, using jigdo format v2
   + Changes to libjte to generate the new format as an option - still
     defaults to v1 for cmpatibility for now
   + libjte updated to major version 2
   + Add support for v2 in jigdump, jigit-mkimage etc. too
   + Add new jigsum-sha256 program, for the base64-like output with
     sha256 checksums
 * jigdump now also understands jigdo .iso.tmp files
 * Misc small cleanups

Version 1.21 (2019-02-04)

 * parallel-sums:
   + New utility for calculating checksums in parallel
 * Misc code cleanups
 * Misc man page cleanups
 * jigit-mkimage:
   + Print missing file names correctly

Version 1.20 (2014-08-18)

 * mkjigsnap:
   + Add support for tracking back-references, i.e. which jigdo did a
     particular file come from.
   * Add support for verifying checksums
 * jigsum:
   * Add support for verifying checksums, similar to md5sum etc.
 * Misc code cleanups:
   * warnings
   * sparse errors
   * Fix memset calls to zero the right amount of memory

Version 1.19 (2011-06-16)

 * Improved error-handling in mkimage and libjte.
 * mkjigsnap rewrite:
   + merge two old implementations
   + switch to perl: cleaner code, much faster snapshot generation
   + added -v (verbose) and -N (dry-run) options
 * Put AC_CONFIG_MACRO_DIR in a better place in the libjte configure
   script.
 * Remove -Werror from the Makefile - gcc is getting more and more
   pedantic with every release, and this may cause package builds to fail.

Version 1.18 (2011-04-30)

 * More dynamic memory handling in mkimage, instead of allocating
   fixed-size strings on the stack. Should now work on Hurd too.
 * Fix man page syntax errors.
 * s,config.h,../config.h, in libjte for easier use by the xorriso folks

Version 1.17 (2011-03-16)

 * Rename mkimage to jigit-mkimage, to avoid a naming clash.
 * Sanity-check jigdo files - require a valid JigsawDownload
   header by default.
 * Multiple fixes for how MD5 comparisons are dealt with.
 * Add libjte - jigdo generation code that used to be embedded in
   mkisofs/genisoimage. Now a shared library useable in xorriso etc.

Version 1.16 (2007-10-18)

 * Remove build-stamp from the release tarball, to make debian package
   builds work better
 * Fix wget options typo
 * Add proper working bzip2-compressed template support in jigdump and
   mkimage
 * In mkimage, don't stat every file mentioned in the md5 list until we
   need to.
 * Other small cleanups in jigdump and mkimage

Version 1.15 (2005-01-17)

 * Minor compilation fixes
 * Cope with missing files better
 * Minor jigdo file-handling cleanups
 * mkimage code refactored to make it easier to maintain
 * Cleaned up licensing - added explicit GPL v2 headers all over
 * Fix image reconstruction with jigit

Version 1.14 (2004-10-12)

 * Fixed generation of rsyncsums
 * Added -d option to mkjigsnap to specify DIRNAME
 * Minor prettiness tweaks to output UI of mkimage and jigit
 * 64-bit printf fixes

Version 1.13 (2004-09-05)

 * Improved the documentation, listing new options to mkimage
 * Fixed/updated the debian-cd patch

Version 1.12 (2004-09-05)

 * Improved the documentation, listing new options to mkimage
 * Fixed/updated the debian-cd patch

Version 1.11 (2004-09-05)

 * Added rsync sums into JTE output
 * Added jigit script
 * Added extra options to mkimage to support jigit

Version 1.10 (2004-09-01)

 * Removed the last md5 check; no longer needed
 * Added new -jigdo-force-md5 option; any files matching specified
   paths MUST match an MD5 supplied, or fail
 * Fixed a couple of typos.
 * Added initial work for bz2 support.
 * debian-cd diff now includes updated grab_md5 and jigdo_cleanup
   scripts that must be used.
 * mkisofs now leaves a blank [Servers] section in the .jigdo file -
   jigdo_cleanup does the work.

Version 1.9 (2004-08-30)

 * Added MD5 reverse lookups to pick up boot files; helps reduce
   CD/DVD #1 size substantially, BUT -md5-list option now mandatory.
 * Fixed sizing bug in HFS hybrid template code.

Version 1.8 (2004-08-27)

 * Tweaked the -exclude option to reduce template file sizes.

Version 1.7 (2004-08-19)

 * Fixed a couple of silly bugs introduced; don't assume template
   creation.
 * Added README files and man page sections for the new boot
   support.

Version 1.6 (2004-07-14)

 * Added code to check MD5 sums in mkisofs; should no longer need
   to run mirror_check any more. See the -md5-list option.

Version 1.5 (2004-07-05)

 * Minor bug-fixes and updates to mkimage
 * Added iso-image.pl wrapper script

Version 1.4 (2004-06-26)

 * Added mkimage
 * Moved utils and patch into a tarball

Version 1.3 (2004-06-18)

 * Fixed crash without jigdo options
 * Cleaned up mipsel boot code
 * Make it easier to specify the boot command line for hppa
 * Minor cosmetic .jigdo cleanup

Version 1.2 (2004-06-14)

 * Added boot support for alpha, hppa, mips, mipsel
 * Cosmetic changes to the .jigdo file

Version 1.1 (2004-06-10)

 * Removed .jte file code
 * Large cleanup to reduce patch size

Version 1.0 (2004-06-07)

 * Initial release
 * Produced .jigdo, .template and .jte files


jigdump.c
/*
 * jigdump
 *
 * Tool to dump the contents of a jigdo template file
 *
 * Copyright (c) 2004-2019 Steve McIntyre <steve@einval.com>
 *
 * GPL v2 - see COPYING
 */

#include <errno.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <zlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>
#include "jig-base64.h"
#include "md5.h"
#include "jigdo.h"

#define HEADER_STRING "JigsawDownload template"

static INT64 find_string(unsigned char *buf, size_t buf_size, char *search)
{
    size_t length = strlen(search);
    INT64 result;
    
    for (result = 0; result < (buf_size - length); result++)
    {
        if (!memcmp(&buf[result], search, length))
            return result;
    }
    return -1;
}

static INT64 parse_data_block(INT64 offset, unsigned char *buf, size_t buf_size)
{
    /* Parse the contents of this data block... */
    UINT64 dataLen = 0;
    UINT64 dataUnc = 0;

    if (!strncmp((char *)buf, "DATA", 4))
        printf("\ngzip data block found at offset %lld\n", offset);
    else
        printf("\nbzip2 data block found at offset %lld\n", offset);

    dataLen = (UINT64)buf[4];
    dataLen |= (UINT64)buf[5] << 8;
    dataLen |= (UINT64)buf[6] << 16;
    dataLen |= (UINT64)buf[7] << 24;
    dataLen |= (UINT64)buf[8] << 32;
    dataLen |= (UINT64)buf[9] << 40;
    printf("  compressed block size %llu bytes\n", dataLen);

    dataUnc = (UINT64)buf[10];
    dataUnc |= (UINT64)buf[11] << 8;
    dataUnc |= (UINT64)buf[12] << 16;
    dataUnc |= (UINT64)buf[13] << 24;
    dataUnc |= (UINT64)buf[14] << 32;
    dataUnc |= (UINT64)buf[15] << 40;
    printf("  uncompressed block size %llu bytes\n", dataUnc);

    return dataLen;
}

static INT64 parse_desc_block(INT64 offset, unsigned char *buf, size_t buf_size)
{
    /* Parse the contents of this data block... */
    UINT64 descLen = 0;
    
    printf("\nDESC block found at offset %lld (%llx)\n", offset, offset);
    descLen = (UINT64)buf[4];
    descLen |= (UINT64)buf[5] << 8;
    descLen |= (UINT64)buf[6] << 16;
    descLen |= (UINT64)buf[7] << 24;
    descLen |= (UINT64)buf[8] << 32;
    descLen |= (UINT64)buf[9] << 40;
    printf("  DESC block size is %llu bytes\n", descLen);
    
    return 10;
}

static INT64 parse_desc_data(INT64 offset, unsigned char *buf, size_t buf_size)
{
    int type = buf[0];
    printf("  DESC entry: block type %d at offset %lld (0x%llx)\n",
	   type, offset, offset);
    
    switch (type)
    {
        case BLOCK_DATA:
        {
            UINT64 skipLen = 0;
            skipLen = (UINT64)buf[1];
            skipLen |= (UINT64)buf[2] << 8;
            skipLen |= (UINT64)buf[3] << 16;
            skipLen |= (UINT64)buf[4] << 24;
            skipLen |= (UINT64)buf[5] << 32;
            skipLen |= (UINT64)buf[6] << 40;
            printf("    7 byte entry:\n");
            printf("      Template data, %llu bytes of data\n", skipLen);
            return 7;
        }
        case BLOCK_IMAGE_MD5:
        {
            UINT64 imglen = 0;
            UINT32 blocklen = 0;
            int i = 0;

            imglen = (UINT64)buf[1];
            imglen |= (UINT64)buf[2] << 8;
            imglen |= (UINT64)buf[3] << 16;
            imglen |= (UINT64)buf[4] << 24;
            imglen |= (UINT64)buf[5] << 32;
            imglen |= (UINT64)buf[6] << 40;

            blocklen = (UINT32)buf[23];
            blocklen |= (UINT32)buf[24] << 8;
            blocklen |= (UINT32)buf[25] << 16;
            blocklen |= (UINT32)buf[26] << 24;

	    printf("    27 byte entry:\n");
	    printf("      Image info block v1 using MD5\n");
            printf("      Original image length %llu bytes\n", imglen);
            printf("      Image MD5: ");
            for (i = 7; i < 23; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[7], 16));
            printf("      Rsync block length %lu bytes\n", blocklen);
            return 27;
        }
        case BLOCK_IMAGE_SHA256:
        {
            UINT64 imglen = 0;
            UINT32 blocklen = 0;
            int i = 0;

            imglen = (UINT64)buf[1];
            imglen |= (UINT64)buf[2] << 8;
            imglen |= (UINT64)buf[3] << 16;
            imglen |= (UINT64)buf[4] << 24;
            imglen |= (UINT64)buf[5] << 32;
            imglen |= (UINT64)buf[6] << 40;

            blocklen = (UINT32)buf[39];
            blocklen |= (UINT32)buf[40] << 8;
            blocklen |= (UINT32)buf[41] << 16;
            blocklen |= (UINT32)buf[42] << 24;

	    printf("    43 byte entry:\n");
	    printf("      Image info block v2 using SHA256\n");
            printf("      Original image length %llu bytes\n", imglen);
            printf("      Image SHA256: ");
            for (i = 7; i < 39; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[7], 32));
            printf("      Rsync block length %lu bytes\n", blocklen);
            return 43;
        }
        case BLOCK_MATCH_MD5:
        case BLOCK_WRITTEN_MD5:
        {
            UINT64 fileLen = 0;
            int i = 0;

            fileLen = (UINT64)buf[1];
            fileLen |= (UINT64)buf[2] << 8;
            fileLen |= (UINT64)buf[3] << 16;
            fileLen |= (UINT64)buf[4] << 24;
            fileLen |= (UINT64)buf[5] << 32;
            fileLen |= (UINT64)buf[6] << 40;
            
	    printf("    31 byte entry:\n");
            printf("      File %s block v1 using MD5\n",
                   (type == BLOCK_MATCH_MD5 ? "match" : "written"));
            printf("      length %llu bytes\n", fileLen);
            printf("      file rsyncsum: ");
            for (i = 7; i < 15; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[7], 8));
            printf("      file md5: ");
            for (i = 15; i < 31; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[15], 16));
            return 31;
        }
        case BLOCK_MATCH_SHA256:
        case BLOCK_WRITTEN_SHA256:
        {
            UINT64 fileLen = 0;
            int i = 0;

            fileLen = (UINT64)buf[1];
            fileLen |= (UINT64)buf[2] << 8;
            fileLen |= (UINT64)buf[3] << 16;
            fileLen |= (UINT64)buf[4] << 24;
            fileLen |= (UINT64)buf[5] << 32;
            fileLen |= (UINT64)buf[6] << 40;

	    printf("    47 byte entry:\n");
            printf("      File %s block v2 using SHA256\n",
                   (type == BLOCK_MATCH_SHA256 ? "match" : "written"));
            printf("      length %llu bytes\n", fileLen);
            printf("      file rsyncsum: ");
            for (i = 7; i < 15; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[7], 8));
            printf("      file sha256: ");
            for (i = 15; i < 47; i++)
                printf("%2.2x", buf[i]);
            printf(" (%s)\n", base64_dump(&buf[15], 32));
            return 47;
        }
        default:
	    printf("    Unrecognised block type %d, skipping forwards in hope\n", type);
	    return 1;
            break;
    }

    return 0;
}

static INT64 parse_desc_pointer(INT64 offset, unsigned char *buf, size_t buf_size, INT64 template_size, INT64 desc_offset)
{
    /* Parse the contents of this final block... */
    UINT64 ptr = 0;

    ptr = (UINT64)buf[0];
    ptr |= (UINT64)buf[1] << 8;
    ptr |= (UINT64)buf[2] << 16;
    ptr |= (UINT64)buf[3] << 24;
    ptr |= (UINT64)buf[4] << 32;
    ptr |= (UINT64)buf[5] << 48;
    printf("  DESC pointer offset %llu\n", ptr);
    printf("  points to DESC at offset %llu\n", template_size - ptr);
    printf("  actual DESC section starts at %llu\n", desc_offset);
    if (template_size - ptr != desc_offset)
	printf("ERROR: pointers don't match up\n");

    return 6;
}

static INT64 parse_tmp_desc_pointer(unsigned char *buf)
{
    /* Parse the contents of this final block... */
    UINT64 ptr = 0;

    ptr = (UINT64)buf[0];
    ptr |= (UINT64)buf[1] << 8;
    ptr |= (UINT64)buf[2] << 16;
    ptr |= (UINT64)buf[3] << 24;
    ptr |= (UINT64)buf[4] << 32;
    ptr |= (UINT64)buf[5] << 48;

    return ptr;
}

static INT64 parse_header_block(INT64 offset, unsigned char *buf, size_t buf_size)
{
    /* Massive overkill for sizes, but safe! */
    char format_version[BUF_SIZE] = {0};
    char generator[BUF_SIZE] = {0};
    int ret = 0;

    ret = sscanf((char *)buf, HEADER_STRING" %s %s", format_version, generator);
    if (ret != 2)
    {
        printf("Failed to parse header, ret %d\n", ret);
	return -1;
    }
    printf("Found header: \"%s\"\n", HEADER_STRING);
    printf("Format version: %s\n", format_version);
    printf("Generator: %s\n", generator);

    return 0;
}

/* return 1 if haystack ends with the text contained in needle, 0
 * otherwise */
static int string_ends(char *haystack, char *needle)
{
    int hlen;
    int nlen;

    if (NULL == haystack || NULL == needle)
        return 0;

    hlen = strlen(haystack);
    nlen = strlen(needle);
    if (nlen > hlen)
        return 0;
    if (!strcmp((haystack + hlen - nlen), needle))
        return 1;

    return 0;
}

int main(int argc, char **argv)
{
    char *filename = NULL;
    int fd = -1;
    unsigned char *buf = NULL;
    INT64 offset = 0;
    INT64 bytes = 0;
    INT64 template_size = 0;
    INT64 desc_offset = 0;
    struct stat sb;
    e_state state = STARTING;
    
    if (argc != 2)
    {
        printf("No filename specified! Try again...\n");
        return EINVAL;
    }
    
    filename = argv[1];
    
    fd = open(filename, O_RDONLY);
    if (-1 == fd)
    {
        printf("Failed to open file %s, error %d!. Try again...\n", filename, errno);
        return errno;
    }

    if (-1 == fstat(fd, &sb))
    {
        printf("Failed to stat file %s, error %d!. Try again...\n", filename, errno);
        return errno;
    }
    template_size = sb.st_size;

    buf = malloc(BUF_SIZE);
    if (!buf)
    {
        printf("Failed to malloc %d bytes. Abort!\n", BUF_SIZE);
        return ENOMEM;
    }

    /* If we're looking at a template file, find the beginning of the
     * data - read the first chunk, including the header. Only do this
     * if the filename ends in ".template" */
    if (string_ends(filename, ".template"))
    {
        printf("Filename %s ends in .template.\n", filename);
        printf("Assuming this is meant to be a template file.\n");
	while (STARTING == state)
	{
	    INT64 start_offset = -1;

	    bytes = read(fd, buf, BUF_SIZE);
	    if (0 >= bytes)
	    {
		state = DONE;
		break;
	    }

	    start_offset = find_string(buf, bytes, HEADER_STRING);
	    if (0 != start_offset)
	    {
		printf("Can't find header - this is not a template file\n");
		return EINVAL;
	    }
	    if (0 != parse_header_block(start_offset, buf, bytes))
		return EINVAL;

	    start_offset = find_string(buf, bytes, "DATA");
	    if (-1 == start_offset)
		start_offset = find_string(buf, bytes, "BZIP");
	    if (start_offset >= 0)
	    {
		offset += start_offset;
		state = IN_DATA;
		break;
	    }
	    offset += bytes;
	}
    }
    else
    {
        printf("Filename %s does not end in .template.\n", filename);
        printf("Assuming this is NOT meant to be a template file.\n");
        /* We're looking at a tmp file? No header available. Seek to
        the end for the final DESC * pointer */
        lseek(fd, template_size - 6, SEEK_SET);
        bytes = read(fd, buf, BUF_SIZE);
        if (bytes != 6)
        {
            state = DONE;
        }
        else
        {
            INT64 start_offset = parse_tmp_desc_pointer(buf);
            offset = template_size - start_offset;
            state = IN_DESC;
        }
    }

    while (DONE != state && ERROR != state)
    {
        INT64 start_offset = -1;
        lseek(fd, offset, SEEK_SET);
        bytes = read(fd, buf, BUF_SIZE);
        if (0 >= bytes)
        {
            state = ERROR;
            break;
        }
        if (IN_DATA == state)
        {
            if (!find_string(buf, bytes, "DATA") || !find_string(buf, bytes, "BZIP"))
                state = IN_DATA;
            if (!find_string(buf, bytes, "DESC"))
	    {
                state = IN_DESC;
		desc_offset = offset;
	    }
        }
        switch (state)
        {
            case IN_DATA:
                start_offset = parse_data_block(offset, buf, bytes);
                offset += start_offset;
                break;
            case IN_DESC:
                start_offset = parse_desc_block(offset, buf, bytes);
                offset += start_offset;
                state = DUMP_DESC;
                break;
            case DUMP_DESC:
                start_offset = parse_desc_data(offset, buf, bytes);
                offset += start_offset;
		if (offset >= (template_size - 6))
			state = DUMP_DESC_PTR;;
                break;
	    case DUMP_DESC_PTR:
		state = DONE;
		start_offset = parse_desc_pointer(offset, buf, bytes, template_size, desc_offset);
		offset += start_offset;
		break;
            default:
                break;
        }
	if (offset == template_size)
	    state = DONE;
    }        
    
    close(fd);

    return 0;
}
